ext/fg/js/frontend-initialize.js

```diff
@@ -25,49 +25,99 @@
+        ({action, params}, {resolve}) => {
+            if (action === 'rootPopupInformation') {
+                resolve(params);
+            }
+        }
+    );
+    apiForward('rootPopupRequestInformationBroadcast');
+    const {popupId, frameId} = await rootPopupInformationPromise;
+
+    const frameOffsetForwarder = new FrameOffsetForwarder();
+    frameOffsetForwarder.start();
+    const getFrameOffset = frameOffsetForwarder.getOffset.bind(frameOffsetForwarder);
+
+    const popup = new PopupProxy(popupId, 0, null, frameId, url, getFrameOffset);
+    await popup.prepare();
+
+    return popup;
+}
+
+async function getOrCreatePopup(depth) {
+    const frameOffsetForwarder = new FrameOffsetForwarder();
+    frameOffsetForwarder.start();
+
+    const popupHost = new PopupProxyHost();
+    await popupHost.prepare();
+
+    const popup = popupHost.getOrCreatePopup(null, null, depth);
+
+    return popup;
+}
+
+async function createPopupProxy(depth, id, parentFrameId, url) {
+    const popup = new PopupProxy(null, depth + 1, id, parentFrameId, url);
+    await popup.prepare();
+
+    return popup;
+}
+
 async function main() {
     await yomichan.prepare();
 
     const data = window.frontendInitializationData || {};
-    const {id, depth=0, parentFrameId, url, proxy=false} = data;
-
-    const optionsContext = {depth, url};
-    const options = await apiOptionsGet(optionsContext);
-
-    let popup;
-    if (!proxy && (window !== window.parent) && options.general.showIframePopupsInRootFrame) {
-        const rootPopupInformationPromise = yomichan.getTemporaryListenerResult(
-            chrome.runtime.onMessage,
-            ({action, params}, {resolve}) => {
-                if (action === 'rootPopupInformation') {
-                    resolve(params);
-                }
-            }
-        );
-        apiForward('rootPopupRequestInformationBroadcast');
-        const {popupId, frameId} = await rootPopupInformationPromise;
-
-        const frameOffsetForwarder = new FrameOffsetForwarder();
-        frameOffsetForwarder.start();
-        const getFrameOffset = frameOffsetForwarder.getOffset.bind(frameOffsetForwarder);
-
-        popup = new PopupProxy(popupId, 0, null, frameId, url, getFrameOffset);
-        await popup.prepare();
-    } else if (proxy) {
-        popup = new PopupProxy(null, depth + 1, id, parentFrameId, url);
-        await popup.prepare();
-    } else {
-        const frameOffsetForwarder = new FrameOffsetForwarder();
-        frameOffsetForwarder.start();
-
-        const popupHost = new PopupProxyHost();
-        await popupHost.prepare();
-
-        popup = popupHost.getOrCreatePopup(null, null, depth);
-    }
-
-    const frontend = new Frontend(popup);
-    await frontend.prepare();
+    const {id, depth=0, parentFrameId, url, proxy=false, isSearchPage=false} = data;
+
+    const isIframe = !proxy && (window !== window.parent);
+
+    const initEventDispatcher = new EventDispatcher();
```

Comment:
The use of `initEventDispatcher ` can probably be removed by some simplification.

`initEventDispatcher` is only used by `frontend`, which is only instantiated once. Add a `setPopup` method on it for `popupChange`, and `setDisabledOverride` can probably be handled in a similar way, or by setting the popup to `null`.

Reply:
https://github.com/FooSoft/yomichan/pull/439/commits/44ff4d889942987adf365a701c01eb8c3f98c2dd

I'd rather not set popup to `null` because then I'd have to potentially handle nulls all the places where popup is used.